### DISCUSSION TITLE
Add next trouble/quickfix and prev trouble/quickfix item keymaps?

### DISCUSSION THREAD

**Discussion by briandipalma** [LazyVim/LazyVim#278] [Ideas]
Would a PR adding `[q` and `]q` to navigate trouble with fallback to quickfix lists be OK? I don't think they'd clash with anything and I think they cut out having to think about window movements to get back into the lists.

**Comment by dpetka2001** [comment]
I'd be interested in seeing your config about this even if it doesn't get merged as a PR.

**Comment by briandipalma** [comment]
```lua
if Util.has("trouble.nvim") then
	vim.keymap.set("n", "[q", function()
		require("trouble").previous({ skip_groups = true, jump = true })
	end, { desc = "Previous trouble item" })
	vim.keymap.set("n", "]q", function()
		require("trouble").next({ skip_groups = true, jump = true })
	end, { desc = "Next trouble item" })
else
	vim.keymap.set("n", "[q", vim.cmd.cprev, { desc = "Previous quickfix" })
	vim.keymap.set("n", "]q", vim.cmd.cnext, { desc = "Next quickfix" })
end
```

The issue here is that I need to check if trouble or quickfix list is open in the trouble functions and execute accordingly. Could you make `is_open` part of trouble's API @folke ? I can raise PRs in both repos if this seems reasonable. I could use `getqflist` too possibly but I feel trouble should be checked ahead of the existence of a quickfix list.

**Reply by folke** [reply]
I've just pushed a change to trouble to expose `require("trouble").is_open()`, but I'm not 100% sure yet if we should include this in LazyVim and how it should work.

The best way would probably be to add the trouble specific keybind as `keys` to `trouble`. In those keybinds, check if  `is_open` and if not execute `vim.cmd.cprev`.

In `lua/lazvim/config/keymaps.lua`, then just have a `if not  Util.has("trouble.nvim")` block to set the quickfix only bindings.

That's probably the best way?

**Reply by folke** [reply]
A PR like that does seem useful, so feel free to give it a go :)

**Reply by briandipalma** [reply]
I was thinking of having the keymap exist in all circumstances. The function that is called will check for trouble's existance and if it's open, if that's the case it will call the trouble callback else it will call the quickfix equivalent. The description will need to be different too based on trouble's existence too. Seems reasonable, I'll raise a PR so there's something concrete to look at.

**Reply by briandipalma** [reply]
PR: https://github.com/LazyVim/LazyVim/pull/299
